### ISSUE TITLE
File-local, directory-local or namespaced reference id's

### ISSUE THREAD

**Issue by danielytics** [entangled/entangled#88]
When a codebase gets large and especially when it gets split into many (literate) files across multiple directories, making sure that reference id's stay unique becomes difficult. This is especially true if two completely unrelated parts of the code use them, perhaps both are even written by different people or at different times. It becomes easy to have reference id's clash and accidentally merge unrelated code together.

The current solution is to either use very specific and long id's, that are unlikely to clash, or to prefix every reference id with some namespace. Neither solution is particularly satisfying as it leads to very long and verbose reference id's, making the literate code harder to read. (Eg consider `<<check-results>>` vs something like `<<some-module/sub-module/check-results>>`)

I saw that an[ idea for namespaces](https://github.com/entangled/entangled/pull/16#issuecomment-575065392) was mentioned. Ideally, these could be set some way that doesn't make the reference id harder to read.

Two potential alternative solutions that spring to mind are:

1. Have two different types of reference id's: file-local and global, where file-local ones only get merged within one file. This would help, but global references would still potentially cause clashes, but that could then be resolved through prefixing (which at least only has to be done for references that are intended to "escape" a file)
2. Have reference id's be scoped to the directory that they're in. I'm unsure how this would work with sub-directories (I assume the ideal would be that they do "escape" into sub-directories, but that also means they should escape upwards to parents too? Maybe only if a parent references them? I'm unsure), but two sibling directories should be completely isolated from each other.

I'm not sure if either of those would suit what entangled is going for.

The intent is to allow splitting into multiple files to continue to work as it does now, as it is a very useful way to organise large bits of code into smaller chunks, yet still allowing unrelated code to be isolated from each other.

The obvious third option is to simply run entangled separately on the different bits of code that should be isolated, which would work right now, but that makes the file watch (`make watch`) solution from the tutorial a lot harder (I assume I'd have to run separate instances of entangled for each separate module), so it doesn't seem particularly satisfying either, but maybe I could be convinced..?

**Comment by jhidding** [comment]
Yes, I've considered namespaces, but so far no one has requested it before, so I didn't prioritize it. Also I haven't taken the time to figure out a precise semantics for this feature. There'd need to be a separator, maybe `::` or `/`, and a way to specify the namespace for a given file, or perhaps the name should be the filename without extension? In the last case we could have:

main.md:
~~~md
The main program:
``` {.python file=main.py}
<<imports>>

if __name__ == "__main__":
    <<hello::say-hello>>
    <<exit-program>>
```

Exit the program:
``` {.python #exit-program}
sys.exit(0)
```
~~~

hello.md:
~~~md
Say hello:
``` {.python #say-hello}
print("Hello, World")
```
~~~

The separator needs to be parsable as part of the code-block id, so that you can say:

imports.md:
~~~md
We need the `sys` module:
``` {.python #main::imports}
import sys
```
~~~

Names without explicit namespaces need to be read as having the current namespace prefixed. Then the internal logic of Entangled can stay as it is.

Let me know what you think about this proposal.

**Comment by danielytics** [comment]
> Names without explicit namespaces need to be read as having the current namespace prefixed. Then the internal logic of Entangled can stay as it is.

I assume this would break existing setups that rely on cross-file references? Its not a problem for me, but maybe others wouldn't like that.

So, the idea is basically that references are local to the file they're in, unless namespaced to reference another file? That would work perfectly well for me. Its kinda like my option 2, except "global" is namespaced and local can still be accessed from elsewhere using the namespaced syntax. :+1:

**Comment by jhidding** [comment]
This would be an entirely optional feature, that has to be enabled in `entangled.dhall`
